include/performance_test_fixture/mimick_memory_manager.hpp

```diff
@@ -0,0 +1,78 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef PERFORMANCE_TEST_FIXTURE__MIMICK_MEMORY_MANAGER_HPP_
+#define PERFORMANCE_TEST_FIXTURE__MIMICK_MEMORY_MANAGER_HPP_
+
+#include <benchmark/benchmark.h>
+
+#include <functional>
+#include <unordered_set>
+#include <memory>
+#include <mutex>
+#include <utility>
+
+// Defined by mimick.h
+struct mmk_stub;
+
+// Defined by mimick_utils.hpp
+template<class T>
+struct mmk_allocator;
+
+namespace performance_test_fixture
+{
+
+class MimickMemoryManager : public benchmark::MemoryManager
+{
+public:
+  MimickMemoryManager();
+
+  ~MimickMemoryManager();
+
+  virtual void Pause();
+
+  virtual void Reset();
+
+  virtual void Resume();
+
+  void Start() override;
+
+  void Stop(benchmark::MemoryManager::Result * result) override;
+
+private:
+  void Stop() noexcept;
+
+  void * on_calloc(size_t nitems, size_t size);
+  void on_free(void * ptr);
+  void * on_malloc(size_t size);
+  void * on_realloc(void * ptr, size_t size);
+
+  std::mutex stat_lock;
+  bool recording_enabled;
+
+  decltype(benchmark::MemoryManager::Result::max_bytes_used) cur_bytes_used;
+  decltype(benchmark::MemoryManager::Result::max_bytes_used) max_bytes_used;
+  decltype(benchmark::MemoryManager::Result::num_allocs) num_allocs;
+  std::unique_ptr<std::unordered_set<void *, std::hash<void *>, std::equal_to<void *>,
+    mmk_allocator<void *>>> ptr_set;
```

Comment:
@cottsay why the extra level of indirection using `std::unique_ptr`?

Reply:
I did that so that I could forward-declare `mmk_allocator` and avoid including "mimick.h" in this public API header.